v1/topdown/graphql.go

```diff
@@ -453,28 +502,143 @@ func builtinGraphQLIsValid(_ BuiltinContext, operands []*ast.Term, iter func(*as
+	// Schemas are only cached if they are valid
+	schemaCacheKey, schema := cacheGetSchema(bctx, operands)
+	if schema == nil {
+		var schemaDoc *gqlast.SchemaDocument
+		var validatedSchema *gqlast.Schema
+
+		switch x := operands[0].Value.(type) {
+		case ast.String:
+			schemaDoc, err = parseSchema(string(x))
+		case ast.Object:
+			schemaDoc, err = objectToSchemaDocument(x)
+		default:
+			// Error if wrong type.
+			return iter(ast.InternedBooleanTerm(false))
+		}
+		if err != nil {
+			return iter(ast.InternedBooleanTerm(false))
+		}
+		// Validate the schema, this determines the result
+		// and whether there is a schema to cache
+		validatedSchema, err = convertSchema(schemaDoc)
+		if bctx.InterQueryBuiltinValueCache != nil && err == nil {
+			cacheInsertSchema(bctx, schemaCacheKey, validatedSchema)
+		}
 	}
 
-	// Validate the schema, this determines the result
-	_, err = convertSchema(schemaDoc)
 	return iter(ast.InternedBooleanTerm(err == nil))
 }
 
+// Insert Schema into cache
+func cacheInsertSchema(bctx BuiltinContext, key string, schema *gqlast.Schema) int {
+	cacheKeyAST := ast.StringTerm(key).Value
+	numDroppedEntries := bctx.InterQueryBuiltinValueCache.Insert(cacheKeyAST, schema)
+	return numDroppedEntries
+}
+
+// Insert SchemaAST into cache
+func cacheInsertSchemaAST(bctx BuiltinContext, key string, schemaAST ast.Value) int {
+	cacheKeyAST := ast.StringTerm(key).Value
+	numDroppedEntries := bctx.InterQueryBuiltinValueCache.Insert(cacheKeyAST, schemaAST)
+	return numDroppedEntries
+}
+
+// Insert SchemaDocument into cache
+func cacheInsertSchemaDoc(bctx BuiltinContext, key string, schemaDoc *gqlast.SchemaDocument) int {
+	cacheKeyAST := ast.StringTerm(key).Value
+	numDroppedEntries := bctx.InterQueryBuiltinValueCache.Insert(cacheKeyAST, schemaDoc)
+	return numDroppedEntries
+}
+
+// Returns the cache key and a Schema if this key already exists in the cache
+func cacheGetSchema(bctx BuiltinContext, operands []*ast.Term) (string, *gqlast.Schema) {
+	var key string
+	var schema *gqlast.Schema
+
+	if k, keyOk := cacheKeyWithPrefix(bctx, operands, "gql_schema"); keyOk {
+		key = k
+		if val, ok := bctx.InterQueryBuiltinValueCache.Get(ast.StringTerm(key).Value); ok {
+			var isSchema bool
+			schema, isSchema = val.(*gqlast.Schema)
+			if !isSchema {
+				return key, nil
+			}
+		}
+	}
+	return key, schema
+}
+
+// Returns the cache key and a SchemaDocument if this key already exists in the cache
+// Note: the SchemaDocument is not a validated Schema
+func cacheGetSchemaDoc(bctx BuiltinContext, operands []*ast.Term) (string, *gqlast.SchemaDocument) {
+	var key string
+	var schemaDoc *gqlast.SchemaDocument
+
+	if k, keyOk := cacheKeyWithPrefix(bctx, operands, "gql_schema_doc"); keyOk {
+		key = k
+		if val, ok := bctx.InterQueryBuiltinValueCache.Get(ast.StringTerm(key).Value); ok {
+			var isSchema bool
+			schemaDoc, isSchema = val.(*gqlast.SchemaDocument)
+			if !isSchema {
+				return key, nil
+			}
+		}
+	}
+	return key, schemaDoc
```

Comment:
Nit: just a bit simplified for readability

```suggestion
	if key, keyOk := cacheKeyWithPrefix(bctx, operands, "gql_schema_ast"); keyOk {
		if val, ok := bctx.InterQueryBuiltinValueCache.Get(ast.StringTerm(key).Value); ok {
			if schemaAST, isAstValue := val.(ast.Value); isAstValue {
				return key, schemaAST
			}
		}
		return key, nil
	}
	return "", nil
```